#John Asencio
## Data Transformation Activity

#B. [MUST] Small Sample of TriMet data


In [ ]:
import pandas as pd

Use the pandas.read_csv() method to read the data into a DataFrame.


In [ ]:
df = pd.read_csv("/content/bc_trip259172515_230215.csv")
df

EVENT_NO_TRIP  EVENT_NO_STOP            OPD_DATE  VEHICLE_ID  METERS  \
0        259172515      259172517  15FEB2023:00:00:00        4223      40   
1        259172515      259172517  15FEB2023:00:00:00        4223      48   
2        259172515      259172517  15FEB2023:00:00:00        4223      57   
3        259172515      259172517  15FEB2023:00:00:00        4223      73   
4        259172515      259172517  15FEB2023:00:00:00        4223     112   
..             ...            ...                 ...         ...     ...   
156      259172515      259172530  15FEB2023:00:00:00        4223    5834   
157      259172515      259172531  15FEB2023:00:00:00        4223    5838   
158      259172515      259172531  15FEB2023:00:00:00        4223    5858   
159      259172515      259172531  15FEB2023:00:00:00        4223    5889   
160      259172515      259172531  15FEB2023:00:00:00        4223    5918   

     ACT_TIME  GPS_LONGITUDE  GPS_LATITUDE  GPS_SATELLITES  GPS_HDOP  
0       20469    -122.648137     45.493082              12       0.7  
1       20474    -122.648240     45.493070              12       0.8  
2       20479    -122.648352     45.493123              12       0.8  
3       20484    -122.648385     45.493262              12       0.7  
4       20489    -122.648347     45.493582              12       0.8  
..        ...            ...           ...             ...       ...  
156     21389    -122.677057     45.528040              11       0.8  
157     21424    -122.677003     45.528037               6       2.0  
158     21429    -122.676765     45.528043               7       1.1  
159     21434    -122.676370     45.528047              10       1.1  
160     21439    -122.675990     45.528065              11       0.8  

[161 rows x 10 columns]

#C. [MUST] Filtering


Some of the columns in our TriMet data are not generally useful for our class project. For example, our contact at TriMet told us that the EVENT_NO_STOP column is not used and can be safely eliminated for any type of analysis of the data.

Use pandas.DataFrame.drop() to filter the EVENT_NO_STOP column.



In [ ]:
df2 = df.drop(columns=['EVENT_NO_STOP'])
#df
df2


EVENT_NO_TRIP            OPD_DATE  VEHICLE_ID  METERS  ACT_TIME  \
0        259172515  15FEB2023:00:00:00        4223      40     20469   
1        259172515  15FEB2023:00:00:00        4223      48     20474   
2        259172515  15FEB2023:00:00:00        4223      57     20479   
3        259172515  15FEB2023:00:00:00        4223      73     20484   
4        259172515  15FEB2023:00:00:00        4223     112     20489   
..             ...                 ...         ...     ...       ...   
156      259172515  15FEB2023:00:00:00        4223    5834     21389   
157      259172515  15FEB2023:00:00:00        4223    5838     21424   
158      259172515  15FEB2023:00:00:00        4223    5858     21429   
159      259172515  15FEB2023:00:00:00        4223    5889     21434   
160      259172515  15FEB2023:00:00:00        4223    5918     21439   

     GPS_LONGITUDE  GPS_LATITUDE  GPS_SATELLITES  GPS_HDOP  
0      -122.648137     45.493082              12       0.7  
1      -122.648240     45.493070              12       0.8  
2      -122.648352     45.493123              12       0.8  
3      -122.648385     45.493262              12       0.7  
4      -122.648347     45.493582              12       0.8  
..             ...           ...             ...       ...  
156    -122.677057     45.528040              11       0.8  
157    -122.677003     45.528037               6       2.0  
158    -122.676765     45.528043               7       1.1  
159    -122.676370     45.528047              10       1.1  
160    -122.675990     45.528065              11       0.8  

[161 rows x 9 columns]


For this in-class assignment we won’t need the GPS_SATELLITES or GPS_HDOP columns, so drop them as well.



In [ ]:
df3 = df2.drop(columns=['GPS_SATELLITES', 'GPS_HDOP'])
df3

EVENT_NO_TRIP            OPD_DATE  VEHICLE_ID  METERS  ACT_TIME  \
0        259172515  15FEB2023:00:00:00        4223      40     20469   
1        259172515  15FEB2023:00:00:00        4223      48     20474   
2        259172515  15FEB2023:00:00:00        4223      57     20479   
3        259172515  15FEB2023:00:00:00        4223      73     20484   
4        259172515  15FEB2023:00:00:00        4223     112     20489   
..             ...                 ...         ...     ...       ...   
156      259172515  15FEB2023:00:00:00        4223    5834     21389   
157      259172515  15FEB2023:00:00:00        4223    5838     21424   
158      259172515  15FEB2023:00:00:00        4223    5858     21429   
159      259172515  15FEB2023:00:00:00        4223    5889     21434   
160      259172515  15FEB2023:00:00:00        4223    5918     21439   

     GPS_LONGITUDE  GPS_LATITUDE  
0      -122.648137     45.493082  
1      -122.648240     45.493070  
2      -122.648352     45.493123  
3      -122.648385     45.493262  
4      -122.648347     45.493582  
..             ...           ...  
156    -122.677057     45.528040  
157    -122.677003     45.528037  
158    -122.676765     45.528043  
159    -122.676370     45.528047  
160    -122.675990     45.528065  

[161 rows x 7 columns]


Next, start over and this time try filtering these same columns using the usecols parameter of the read_csv() method.  

Why might we want to filter columns this way instead of using drop()?

In [ ]:
use_df = pd.read_csv("/content/bc_trip259172515_230215.csv", usecols=['EVENT_NO_TRIP','OPD_DATE','VEHICLE_ID','METERS','ACT_TIME','GPS_LONGITUDE','GPS_LATITUDE'])
use_df

EVENT_NO_TRIP            OPD_DATE  VEHICLE_ID  METERS  ACT_TIME  \
0        259172515  15FEB2023:00:00:00        4223      40     20469   
1        259172515  15FEB2023:00:00:00        4223      48     20474   
2        259172515  15FEB2023:00:00:00        4223      57     20479   
3        259172515  15FEB2023:00:00:00        4223      73     20484   
4        259172515  15FEB2023:00:00:00        4223     112     20489   
..             ...                 ...         ...     ...       ...   
156      259172515  15FEB2023:00:00:00        4223    5834     21389   
157      259172515  15FEB2023:00:00:00        4223    5838     21424   
158      259172515  15FEB2023:00:00:00        4223    5858     21429   
159      259172515  15FEB2023:00:00:00        4223    5889     21434   
160      259172515  15FEB2023:00:00:00        4223    5918     21439   

     GPS_LONGITUDE  GPS_LATITUDE  
0      -122.648137     45.493082  
1      -122.648240     45.493070  
2      -122.648352     45.493123  
3      -122.648385     45.493262  
4      -122.648347     45.493582  
..             ...           ...  
156    -122.677057     45.528040  
157    -122.677003     45.528037  
158    -122.676765     45.528043  
159    -122.676370     45.528047  
160    -122.675990     45.528065  

[161 rows x 7 columns]

Using the parameter usecols is advantageous because you can read into only a subset of the data instead of using resources to read in the entire set of data then dropping the columns you don't need.

#D. [MUST] Decoding
Notice that the timestamp for each bread crumb record is encoded in an odd way that might make analysis difficult. The breadcrumb timestamps are represented by two columns, OPD_DATE and ACT_TIME. OPD_DATE merely represents the date on which the bus ran, and it should be constant, unchanging for all breadcrumb records for a single day. The ACT_TIME field indicates an offset, specifically the number of seconds elapsed since midnight on that day.

We’re not sure why TriMet represents the breadcrumb timestamps this way. We do know that this encoding of the timestamps makes automated analysis difficult. So your job is to decode TriMet’s representation and create a new “TIMESTAMP” column containing a pandas.Timestamp value for each breadcrumb.

Suggestions:
Use DataFrame.apply() to apply a function to all rows of your DataFrame
The applied function should input the two to-be-decoded columns, then it should:
create a datetime value from the OPD_DATE input using datetime.strptime()
create a timedelta value from the ACT_TIME
add the timedelta value to the datetime value to produce the resulting timestamp


In [ ]:
from datetime import datetime, timedelta

In [ ]:

def decode_timestamp(opd_date, act_time):
    opd_date = pd.to_datetime(opd_date,format='%d%b%Y:%H:%M:%S')
    timestamp = opd_date + pd.to_timedelta(act_time, unit='s')
    return timestamp

use_df['TIMESTAMP'] = use_df.apply(lambda row: decode_timestamp(row['OPD_DATE'], row['ACT_TIME']), axis=1)

use_df

EVENT_NO_TRIP            OPD_DATE  VEHICLE_ID  METERS  ACT_TIME  \
0        259172515  15FEB2023:00:00:00        4223      40     20469   
1        259172515  15FEB2023:00:00:00        4223      48     20474   
2        259172515  15FEB2023:00:00:00        4223      57     20479   
3        259172515  15FEB2023:00:00:00        4223      73     20484   
4        259172515  15FEB2023:00:00:00        4223     112     20489   
..             ...                 ...         ...     ...       ...   
156      259172515  15FEB2023:00:00:00        4223    5834     21389   
157      259172515  15FEB2023:00:00:00        4223    5838     21424   
158      259172515  15FEB2023:00:00:00        4223    5858     21429   
159      259172515  15FEB2023:00:00:00        4223    5889     21434   
160      259172515  15FEB2023:00:00:00        4223    5918     21439   

     GPS_LONGITUDE  GPS_LATITUDE           TIMESTAMP  
0      -122.648137     45.493082 2023-02-15 05:41:09  
1      -122.648240     45.493070 2023-02-15 05:41:14  
2      -122.648352     45.493123 2023-02-15 05:41:19  
3      -122.648385     45.493262 2023-02-15 05:41:24  
4      -122.648347     45.493582 2023-02-15 05:41:29  
..             ...           ...                 ...  
156    -122.677057     45.528040 2023-02-15 05:56:29  
157    -122.677003     45.528037 2023-02-15 05:57:04  
158    -122.676765     45.528043 2023-02-15 05:57:09  
159    -122.676370     45.528047 2023-02-15 05:57:14  
160    -122.675990     45.528065 2023-02-15 05:57:19  

[161 rows x 8 columns]

#E. [MUST] More Filtering


Now that you have decoded the timestamp you no longer need the OPD_DATE and ACT_TIME columns. Delete them from the DataFrame.

In [ ]:
filter_df = use_df.drop(columns=['OPD_DATE', 'ACT_TIME'])
filter_df

EVENT_NO_TRIP  VEHICLE_ID  METERS  GPS_LONGITUDE  GPS_LATITUDE  \
0        259172515        4223      40    -122.648137     45.493082   
1        259172515        4223      48    -122.648240     45.493070   
2        259172515        4223      57    -122.648352     45.493123   
3        259172515        4223      73    -122.648385     45.493262   
4        259172515        4223     112    -122.648347     45.493582   
..             ...         ...     ...            ...           ...   
156      259172515        4223    5834    -122.677057     45.528040   
157      259172515        4223    5838    -122.677003     45.528037   
158      259172515        4223    5858    -122.676765     45.528043   
159      259172515        4223    5889    -122.676370     45.528047   
160      259172515        4223    5918    -122.675990     45.528065   

              TIMESTAMP  
0   2023-02-15 05:41:09  
1   2023-02-15 05:41:14  
2   2023-02-15 05:41:19  
3   2023-02-15 05:41:24  
4   2023-02-15 05:41:29  
..                  ...  
156 2023-02-15 05:56:29  
157 2023-02-15 05:57:04  
158 2023-02-15 05:57:09  
159 2023-02-15 05:57:14  
160 2023-02-15 05:57:19  

[161 rows x 6 columns]

#F. [MUST] Enhance


Create a new column, called SPEED, that is a calculation of meters traveled per second. Calculate SPEED for each breadcrumb using the breadcrumb’s METERS and TIMESTAMP values along with the METERS and TIMESTAMP values for the immediately preceding breadcrumb record.

Utilize the pandas.DataFrame.diff() method for this calculation. diff() allows you to calculate the difference between a cell value and the preceding row’s value for that same column. Use diff() to create a new dMETERS column and then again to create a new dTIMESTAMP column. Then use apply() (with a lambda function) to calculate SPEED = dMETERS / dTIMESTAMP.
Finally, drop the unneeded dMETERS And dTIMESTAMP columns.



This was a way I figured how to do it without creating the new 'dMETERS' and 'dTIMESTAMP' columns, looks like it gave us the same results

In [ ]:
filter_df['SPEED'] = filter_df['METERS'].diff() / filter_df['TIMESTAMP'].diff().dt.total_seconds()
filter_df

EVENT_NO_TRIP  VEHICLE_ID  METERS  GPS_LONGITUDE  GPS_LATITUDE  \
0        259172515        4223      40    -122.648137     45.493082   
1        259172515        4223      48    -122.648240     45.493070   
2        259172515        4223      57    -122.648352     45.493123   
3        259172515        4223      73    -122.648385     45.493262   
4        259172515        4223     112    -122.648347     45.493582   
..             ...         ...     ...            ...           ...   
156      259172515        4223    5834    -122.677057     45.528040   
157      259172515        4223    5838    -122.677003     45.528037   
158      259172515        4223    5858    -122.676765     45.528043   
159      259172515        4223    5889    -122.676370     45.528047   
160      259172515        4223    5918    -122.675990     45.528065   

              TIMESTAMP     SPEED  
0   2023-02-15 05:41:09       NaN  
1   2023-02-15 05:41:14  1.600000  
2   2023-02-15 05:41:19  1.800000  
3   2023-02-15 05:41:24  3.200000  
4   2023-02-15 05:41:29  7.800000  
..                  ...       ...  
156 2023-02-15 05:56:29  2.800000  
157 2023-02-15 05:57:04  0.114286  
158 2023-02-15 05:57:09  4.000000  
159 2023-02-15 05:57:14  6.200000  
160 2023-02-15 05:57:19  5.800000  

[161 rows x 7 columns]

This is the method using the apply with lambda function that creates the 'dMETERS' and 'dTIMESTAMP' columns.

In [ ]:

filter_df['dMETERS'] = filter_df['METERS'].diff()
filter_df['dTIMESTAMP'] = filter_df['TIMESTAMP'].diff()

filter_df['SPEED'] = filter_df.apply(lambda row: row['dMETERS'] / row['dTIMESTAMP'].total_seconds(), axis=1)

edf = filter_df.drop(columns=['dMETERS', 'dTIMESTAMP'])

edf

EVENT_NO_TRIP  VEHICLE_ID  METERS  GPS_LONGITUDE  GPS_LATITUDE  \
0        259172515        4223      40    -122.648137     45.493082   
1        259172515        4223      48    -122.648240     45.493070   
2        259172515        4223      57    -122.648352     45.493123   
3        259172515        4223      73    -122.648385     45.493262   
4        259172515        4223     112    -122.648347     45.493582   
..             ...         ...     ...            ...           ...   
156      259172515        4223    5834    -122.677057     45.528040   
157      259172515        4223    5838    -122.677003     45.528037   
158      259172515        4223    5858    -122.676765     45.528043   
159      259172515        4223    5889    -122.676370     45.528047   
160      259172515        4223    5918    -122.675990     45.528065   

              TIMESTAMP     SPEED  
0   2023-02-15 05:41:09       NaN  
1   2023-02-15 05:41:14  1.600000  
2   2023-02-15 05:41:19  1.800000  
3   2023-02-15 05:41:24  3.200000  
4   2023-02-15 05:41:29  7.800000  
..                  ...       ...  
156 2023-02-15 05:56:29  2.800000  
157 2023-02-15 05:57:04  0.114286  
158 2023-02-15 05:57:09  4.000000  
159 2023-02-15 05:57:14  6.200000  
160 2023-02-15 05:57:19  5.800000  

[161 rows x 7 columns]

**Question:** What is the minimum, maximum and average speed for this bus on this trip? (Suggestion: use the Dataframe.describe() method to find these statistics)

These are in kilometers/hour

In [ ]:
speed_stats = edf['SPEED'].describe()

print("Minimum speed:", speed_stats['min'])
print("Maximum speed:", speed_stats['max'])
print("Average speed:", speed_stats['mean'])

Minimum speed: 0.0
Maximum speed: 17.4
Average speed: 7.227205815018314


#G. [SHOULD] Larger Data Set
Here is breadcrumb data for the same bus TriMet for the entire day (February 15, 2023): bc_veh4223_230215.csv

Do the same transformations (parts C through F) for this larger data set. Be careful, you might need to treat each trip separately. For example, you might need to find all of the unique values for the EVENT_NO_TRIP column and then do the transformations separately on each trip.



In [2]:
import pandas as pd
df = pd.read_csv("/content/bc_veh4223_230215.csv")
df

EVENT_NO_TRIP  EVENT_NO_STOP            OPD_DATE  VEHICLE_ID  METERS  \
0         259173279      259173350  15FEB2023:00:00:00        4223  194493   
1         259173279      259173350  15FEB2023:00:00:00        4223  194525   
2         259173279      259173350  15FEB2023:00:00:00        4223  194562   
3         259173279      259173350  15FEB2023:00:00:00        4223  194598   
4         259173279      259173350  15FEB2023:00:00:00        4223  194628   
...             ...            ...                 ...         ...     ...   
7401      259173498      259173583  15FEB2023:00:00:00        4223  251109   
7402      259173498      259173583  15FEB2023:00:00:00        4223  251162   
7403      259173498      259173583  15FEB2023:00:00:00        4223  251199   
7404      259173498      259173583  15FEB2023:00:00:00        4223  251238   
7405      259173498      259173583  15FEB2023:00:00:00        4223  251289   

      ACT_TIME  GPS_LONGITUDE  GPS_LATITUDE  GPS_SATELLITES  GPS_HDOP  
0        71279    -122.679572     45.517122               8       1.6  
1        71284    -122.679408     45.517392               8       1.9  
2        71290    -122.679227     45.517700               8       2.2  
3        71295    -122.679048     45.517998               9       0.9  
4        71300    -122.678888     45.518255               9       1.3  
...        ...            ...           ...             ...       ...  
7401     84315    -122.641148     45.539688              12       0.8  
7402     84320    -122.641138     45.540175              12       0.7  
7403     84325    -122.641107     45.540512              12       0.7  
7404     84330    -122.641142     45.540860              12       0.7  
7405     84335    -122.641140     45.541325              12       0.7  

[7406 rows x 10 columns]

###Filtering

In [3]:
df2 = df.drop(columns=['EVENT_NO_STOP', 'GPS_SATELLITES', 'GPS_HDOP'])
#df
df2

EVENT_NO_TRIP            OPD_DATE  VEHICLE_ID  METERS  ACT_TIME  \
0         259173279  15FEB2023:00:00:00        4223  194493     71279   
1         259173279  15FEB2023:00:00:00        4223  194525     71284   
2         259173279  15FEB2023:00:00:00        4223  194562     71290   
3         259173279  15FEB2023:00:00:00        4223  194598     71295   
4         259173279  15FEB2023:00:00:00        4223  194628     71300   
...             ...                 ...         ...     ...       ...   
7401      259173498  15FEB2023:00:00:00        4223  251109     84315   
7402      259173498  15FEB2023:00:00:00        4223  251162     84320   
7403      259173498  15FEB2023:00:00:00        4223  251199     84325   
7404      259173498  15FEB2023:00:00:00        4223  251238     84330   
7405      259173498  15FEB2023:00:00:00        4223  251289     84335   

      GPS_LONGITUDE  GPS_LATITUDE  
0       -122.679572     45.517122  
1       -122.679408     45.517392  
2       -122.679227     45.517700  
3       -122.679048     45.517998  
4       -122.678888     45.518255  
...             ...           ...  
7401    -122.641148     45.539688  
7402    -122.641138     45.540175  
7403    -122.641107     45.540512  
7404    -122.641142     45.540860  
7405    -122.641140     45.541325  

[7406 rows x 7 columns]

###Decoding and filtering

In [5]:
from datetime import datetime, timedelta
def decode_timestamp(opd_date, act_time):
    opd_date = pd.to_datetime(opd_date,format='%d%b%Y:%H:%M:%S')
    timestamp = opd_date + pd.to_timedelta(act_time, unit='s')
    return timestamp

df2['TIMESTAMP'] = df2.apply(lambda row: decode_timestamp(row['OPD_DATE'], row['ACT_TIME']), axis=1)

filter_df = df2.drop(columns=['OPD_DATE', 'ACT_TIME'])
filter_df

EVENT_NO_TRIP  VEHICLE_ID  METERS  GPS_LONGITUDE  GPS_LATITUDE  \
0         259173279        4223  194493    -122.679572     45.517122   
1         259173279        4223  194525    -122.679408     45.517392   
2         259173279        4223  194562    -122.679227     45.517700   
3         259173279        4223  194598    -122.679048     45.517998   
4         259173279        4223  194628    -122.678888     45.518255   
...             ...         ...     ...            ...           ...   
7401      259173498        4223  251109    -122.641148     45.539688   
7402      259173498        4223  251162    -122.641138     45.540175   
7403      259173498        4223  251199    -122.641107     45.540512   
7404      259173498        4223  251238    -122.641142     45.540860   
7405      259173498        4223  251289    -122.641140     45.541325   

               TIMESTAMP  
0    2023-02-15 19:47:59  
1    2023-02-15 19:48:04  
2    2023-02-15 19:48:10  
3    2023-02-15 19:48:15  
4    2023-02-15 19:48:20  
...                  ...  
7401 2023-02-15 23:25:15  
7402 2023-02-15 23:25:20  
7403 2023-02-15 23:25:25  
7404 2023-02-15 23:25:30  
7405 2023-02-15 23:25:35  

[7406 rows x 6 columns]

Questions:
What was the maximum speed for vehicle #4223 on February 15, 2023?

Where and when did this maximum speed occur?

What was the median speed for this vehicle on this day?